modules/core/src/main/java/org/apache/synapse/aspects/flow/statistics/opentracing/management/parentresolving/MessageFlowRepresentationBasedParentResolver.java

```diff
@@ -0,0 +1,87 @@
+ *  under the License.
+ */
+
+package org.apache.synapse.aspects.flow.statistics.opentracing.management.parentresolving;
+
+import org.apache.synapse.aspects.flow.statistics.data.raw.StatisticDataUnit;
+import org.apache.synapse.aspects.flow.statistics.opentracing.management.helpers.TracingUtils;
+import org.apache.synapse.aspects.flow.statistics.opentracing.stores.SpanStore;
+import org.apache.synapse.aspects.flow.statistics.opentracing.models.SpanWrapper;
+
+/**
+ * Resolves the parent based on message flow index, that is explicitly contained by the statistic data unit.
+ */
+public class MessageFlowRepresentationBasedParentResolver extends AbstractParentResolver {
+    /**
+     * Resolves the parent span wrapper based on the parent index, that is explicitly reported by the child statistic
+     * data unit.
+     * @param child     Child statistic data unit.
+     * @param spanStore The span store object.
+     * @return          Resolved parent span wrapper.
+     */
+    public static SpanWrapper resolveParent(StatisticDataUnit child, SpanStore spanStore) {
+        String parentId = String.valueOf(child.getParentIndex());
+        SpanWrapper parent = spanStore.getSpanWrapper(parentId);
+        if (parent != null) {
+            if (isEndpointOrInboundEndpoint(child)) {
+                /*
+                An endpoint can be only parented by either a Call mediator or a Send mediator.
+                 */
+                if (isCallMediator(parent.getStatisticDataUnit()) || isSendMediator(parent.getStatisticDataUnit())) {
+                    return parent;
+                }
+                /*
+                Else:
+                The parent will be chosen by the latest active parent resolver.
+                Endpoints won't have information in artifact holder structuring element stack.
+                 */
+                return LatestActiveParentResolver.resolveParentForEndpointOrInboundEndpoint(spanStore);
+            }
+            if (TracingUtils.isAnonymousSequence(parent.getStatisticDataUnit()) ||
+                    TracingUtils.isAnonymousSequence(child)) {
+                if (isFlowContinuableMediator(parent.getStatisticDataUnit()) ||
+                        isForeachMediator(parent.getStatisticDataUnit())) {
+                    return parent;
+                }
+                return getLatestEligibleParent(spanStore);
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Gets the latest eligible span wrapper,
+     * which is either a flow continuable mediator, or a For Each mediator.
+     * Unlike in the LatestActiveParentResolver, the resolved span wrapper doesn't have to be active.
+     * @param spanStore Span store object.
+     * @return          Resolved parent span wrapper object.
+     */
+    private static SpanWrapper getLatestEligibleParent(SpanStore spanStore) {
+        Object[] spanWrapperKeys = spanStore.getSpanWrappers().keySet().toArray();
```

Comment:
there can be NPEs?

Reply:
No.

`spanStore` is passed down from [this](https://github.com/wso2/wso2-synapse/pull/1466/files#diff-2a3b98173b41bb95122e846bf31228b2R101) point.

When we are initializing a `TracingScope` object [here](https://github.com/wso2/wso2-synapse/pull/1466/files#diff-2a3b98173b41bb95122e846bf31228b2R98), a new `SpanStore` object is created [like this](https://github.com/wso2/wso2-synapse/pull/1466/files#diff-6fec681eedf33528947be8853b7feaf5R62).